documentation/design-docs/ipc-protocol.md

```diff
@@ -844,6 +845,46 @@ struct Payload
     LPCWSTR ClrProductVersion;
 }
 ```
+> Available since .NET 8.0
+
+### `ProcessInfo3`
+
+Command Code: `0x0405`
+
+The `ProcessInfo3` command queries the runtime for some basic information about the process. The returned payload is a variable sized list of key value pairs that provide information about the process.
+
+In the event of an [error](#Errors), the runtime will attempt to send an error message and subsequently close the connection.
+
+#### Inputs:
+
+Header: `{ Magic; Size; 0x0405; 0x0000 }`
+
+There is no payload.
+
+#### Returns (as an IPC Message Payload):
+
+Header: `{ Magic; size; 0xFF00; 0x0000; }`
+
+Payload:
+* `int count`: the number of key value pairs contained in entries
+* `ProcessInfoKeyValuePair[] entries`: the list of key value pairs that provide information about the process
+
+##### Details:
```

Comment:
Do we expect to ever remove fields from this list, or can it be strictly add-only (I'm guessing removal is not that important)? If we can be add-only then perhaps we can append more fields onto the structure defined in GetProcessInfo2 and reuse that command rather than defining a new one? My assumption is that old readers will stop reading the structure after they parse the set of fields currently shown for GetProcessInfo2, ignoring any extra data that is present. Newer readers would parse that data. If there are readers which explicitly check that there is no extra data in GetProcessInfo2 then we could create a ProcessInfo3 command that is very similar and document that we do plan to extend the struct in the future so readers need to ignore extra data on the ProcessInfo3 variation.

I'm a little hesitant to introduce an intermediate layer of abstraction, but if we do then we basically need to describe all the serialization rules we are using to encode it. For example what is the exact name of every supported key (its fine if the list grows over time), what is the meaning of the data published for those keys, and what are all the string formatting rules we are using for every data type that gets serialized as a value. For example is integer 10 = "10", "a", "0xa", "0XA" or "0x0000000A"? Then we'd need to do the same for every other data type that is encoded. It might easier to define a union of different types encoded in the binary formats we have already described serialization rules for rather than defining and implementing string conversions for every type.